### DISCUSSION TITLE
Using LiteX-CNC for 3 axis closed loop servo drive control

### DISCUSSION THREAD

**Discussion by Jaibadon** [Peter-van-Tol/LiteX-CNC#106] [Q&A]
Hi all.
I'm exploring using LiteX-CNC for a project that requires 3-axis closed loop PWM servo drive control(not step/dir). However, I noticed that the LiteX documentation mentions that the PWM module only supports one output. Is there any possibility to use LiteX with more than one PWM output or is this a limitation that cannot be overcome?
Thanks,
Jai.

**Comment by Peter-van-Tol** [comment]
Thank you for considering LitexCNC. 

Short answer: multiple PWM outputs is supported.

This might be wrong use of words in thee documentation. Each module (i.e. PWM) can be added only once to the config, but do support multiple instances or output pins. Up to 32 PWM outputs are supported and tested, but in reality it is only limited by the pins on your board.

If you need help for whipping up your config, please let me know.

**Comment by Jaibadon** [comment]
Hi again!
Just got two 5A-75E's a couple days ago. Turns out I do need some help with the config, just would like some help with some issues I have run into.
When I go to setup PWM like the documentation suggests:
![image](https://github.com/user-attachments/assets/2e9e376b-d986-4cb2-a0aa-21a023e8ce3b)
The compiler returns that it wants me to assign a value to "pin":
![image](https://github.com/user-attachments/assets/799f0ed5-b09a-495d-abfd-d47ed81002a0)

Another issue is when flashing the firmware(with default 5a-75e file) there seems to be a problem with the conversion from svf to bit, svf to flash and bit to flash. 

Flashing the .bit file works just fine:
![image](https://github.com/user-attachments/assets/b36efe5c-224d-4fbc-977c-1b57d68883df)

But when flashed with --permanant or a svf file, the file and board seems to mismatch.
![image](https://github.com/user-attachments/assets/b6da110a-e446-4ccf-8a60-265c22aed8ce)

This isn't an issue with openocd, I have used openFPGALoader with the bit file and got the same hardware and file mismatch, and I don't think it is a hardware issue as this happens on both of my 5A-75E(v8.2) cards plus I have triple checked my wiring.

Any help with these problems would be greatly appreciated.

Thanks,
Jai.

**Reply by Jaibadon** [reply]
Having a glance at the py file, it seems the svf command returns no error when given a bit file and just exits, so the LitexCNC flasher thinks it succeeded?

**Reply by Peter-van-Tol** [reply]
About `PWM`:
- the property `pins` if a the new style, not currently released yet. To use this style, you have to install Litex-CNC from source: `pip install git+https://github.com/Peter-van-Tol/LiteX-CNC`
- when using the original firmware from PyPi, one have to used the `"pin": "j#:#"`

Apparently the documentation is ahead from the release...

**Reply by Peter-van-Tol** [reply]
Can you share your gateware folder, then I can check the contents of the different files with my results. With Yosys, you should be able to flash the `.svf`-file. The `.bit`-file is, after conversion, used to permanently flash to the FPGA. It really surprises me that no error is thrown when trying to flash the `.bit`-file. 

I have a permanently wired combination of a RPi and a FPGA< so it no issue to test it for you...

**Reply by Jaibadon** [reply]
Thanks so much for the help, turns out the flashing issue was a problem with my rPi header extension board. However I am unable to get a response from the board from "ping 10.0.0.10". Does this configuration look right to you?
 
![image](https://github.com/user-attachments/assets/f3878398-cc8a-4414-a68a-e1a7b3abd467)

**Reply by Peter-van-Tol** [reply]
If you're running LinuxCNC on your RPi, I would recommend to use the SPI connection. It's faster and more reliable than the communication over ethernet / UDP.

The connections are looking OK to me, because the wired connection is set to the same range as `10.0.0.10`, but I'm no expert in these connections.

Can you share your config? Things I'm thinking of:
- wrong IP-address, some examples use the `192.168.###.###`-range;
- did you set the`rx-delay`, this should be `0` (will be the default value in the next release);
- can you start the RPi without the Ethernet-cable. When fully booted, insert the ethernet cable and try to `ping` it again. Sometimes a packet is send to the FPGA at start up, which it cannot digest. When this happens, the FPGA can freeze and stop responding.
